hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/Formatter.java

```diff
@@ -0,0 +1,791 @@
+         *      When the exponent is omitted, the floating-point of the number is aligned to the
+         *      decimal separator in the pattern, so there may be overflow/underflow. When the
+         *      exponent is printed, the number is left-aligned to the pattern, and the exponent is
+         *      updated accordingly. </ol>
+         */
+        @Override
+        public String apply(Object input, Locale locale) {
+            DecimalFormatSymbols symbols = DecimalFormatSymbols.getInstance(locale);
+            StringBuilder s = new StringBuilder();
+            String value = input.toString();
+            boolean negative = value.startsWith("-");
+
+            if (value.equals("NaN") || value.endsWith("Infinity")) {
+                // Value is not formattable; pattern is filled with #'s.
+                integerMask.format(s, negative, null, null, 0, symbols);
+                fractionMask.format(s, negative, null, null, 0, symbols);
+            } else {
+                int t = negative ? 1 : 0, begin = value.charAt(t) == '0' ? t + 1 : t;
+                int dot = value.indexOf('.'), exp = dot == -1 ? -1 : value.indexOf('E', dot + 2);
+                String integer = value.substring(begin, dot != -1 ? dot : value.length());
+                String fraction = (dot == -1 ? "" : value.substring(dot + 1, exp != -1 ? exp : value.length()));
+                int exponent = exp == -1 ? 0 : Integer.parseInt(value.substring(exp + 1));
+                // Step 1 - Normalized form
+                String digits = integer + fraction;
+                exponent -= fraction.length();
+                while (digits.startsWith("0")) {
+                    digits = digits.substring(1);
+                }
+                // Step 2 - Find the actual number
+                exponent += this.exponent;
+                // Step 3 - Determine the lengths of the integer and fraction
+                boolean exponential = form == Form.Exponential;
+                int integerLength = exponential ? integerMask.digits : digits.length() + exponent;
+                int fractionLength = exponential ? fractionMask.digits : -exponent;
+                exponent += digits.length() - integerLength;
+
+                // Step 4 - Rounding
+                // integerLength + fractionMask.digits < 0 -> floating-point overflows
+                if ((exponential ? digits.length() > integerLength + fractionMask.digits
+                            : (fractionLength > fractionMask.digits && integerLength + fractionMask.digits >= 0))
+                        && digits.charAt(integerLength + fractionMask.digits) >= '5') {
+                    StringBuilder r = new StringBuilder(digits);
+                    for (int i = integerLength + fractionMask.digits - 1; i >= 0; i--) {
+                        if (r.charAt(i) == '9') {
+                            r.setCharAt(i, '0');
```

Comment:
This line is never hit by the `FormatterTest`, and I wonder about its purpose. We should add a test that covers it.